### PR TITLE
Don't throw on cancellation in chained command handlers

### LINKED ISSUE #52382: When cancellation is requested in formatter, double semicolons are inserted
Repro:
1. Paste the following code into a C# file:
```C#
class C
{
    public string Foo = "foo"
}
```
2. Put a breakpoint [here](https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/CSharp/Formatting/CSharpEditorFormattingService.cs#L176) at the beginning of GetFormattingChangesAsync.
3. Insert semicolon after `"foo"`.
4. After the breakpoint is hit, step over. Hit "Cancel" when this dialog pops up:
image
5. Continue with code execution.
6. Observe double semicolons being inserted:
image

cc @uniqueiniquity

### PR DESCRIPTION
**Pull request by allisonchou**
According to [these Editor API guidelines](https://github.com/Microsoft/vs-editor-api/wiki/Modern-Editor-Commanding-API-Revisited#command-handler-rules-of-handling-a-cancellation-request) for command handlers, we should return early when cancellation is requested instead of throwing an exception, specifically in handlers that may produce side effects upon cancellation.
For example, in Roslyn's format handler, `nextHandler()` is called which triggers the character insertion into the buffer. If cancellation is thrown later on in the format handler, these changes either need to reverted or we need to return early, otherwise the character might be inserted into the buffer twice.

I think I addressed most command handlers that were doing the wrong thing, but let me know if I missed any.
 
Fixes #52382.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
This is super weird :-/

This is the part that feels very underspecified:

>  In other words, the command handler service host must be able to recover to a known consistent state after cancelling your work,

What defines a 'known consistent state'.  It sounds like it's saying: if you bubble cancellation up, you can't have mutated anything on the UI at all.

If that's a requirement, i'd say we need to go further here and we must make all handlers go through a path that simply does not allow cancellation to ever bleed out (as 'no mutation on cancellation' has never been a metaphor we've supported for handlers).

**Comment by CyrusNajmabadi** [comment]
Tagging @jasonmalinowski

File: src/EditorFeatures/CSharp/FixInterpolatedVerbatimString/FixInterpolatedVerbatimStringCommandHandler.cs [hunk 1]
```diff
@@ -35,6 +35,20 @@ public FixInterpolatedVerbatimStringCommandHandler()
         public string DisplayName => CSharpEditorResources.Fix_interpolated_verbatim_string;
 
         public void ExecuteCommand(TypeCharCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            try
```
Comment (sharwell) [inline src/EditorFeatures/CSharp/FixInterpolatedVerbatimString/FixInterpolatedVerbatimStringCommandHandler.cs:39]: 📝 The call to `nextCommandHandler` itself should _not_ be inside the `try` block

File: src/EditorFeatures/CSharp/FixInterpolatedVerbatimString/FixInterpolatedVerbatimStringCommandHandler.cs [hunk 2]
```diff
@@ -35,6 +35,20 @@ public FixInterpolatedVerbatimStringCommandHandler()
...
+                // According to Editor command handler API guidelines, it's best if we return early if cancellation
+                // is requested instead of throwing. Otherwise, we could end up in an invalid state due to already
+                // calling nextCommandHandler().
+            }
+        }
+
+        private static void ExecuteCommandWorker(TypeCharCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
         {
             // We need to check for the token *after* the opening quote is typed, so defer to the editor first
             nextCommandHandler();
```
Comment (sharwell) [inline src/EditorFeatures/CSharp/FixInterpolatedVerbatimString/FixInterpolatedVerbatimStringCommandHandler.cs:54]: 📝 Immediately after this call, the code should check `executionContext.OperationContext.UserCancellationToken`. If cancellation is requested but `nextCommandHandler()` didn't throw, we should just return immediately.

**Comment by allisonchou** [comment]
> If that's a requirement, i'd say we need to go further here and we must make all handlers go through a path that simply does not allow cancellation to ever bleed out (as 'no mutation on cancellation' has never been a metaphor we've supported for handlers).

Most handlers already do the right thing today, there's just a couple that have fallen through the cracks. Furthermore, after addressing Sam's feedback, it seems like we can't just take a "never let cancellation bleed out" approach, as different handlers will need to catch the exception at different locations.

> If they do feel that things should rollback, i would recommend that they themslves start a transaction and they rollback to the starting point so that everything returns back and there doesn't need to be this handshake between us :(

I agree that this would be best for long-term, however I feel like this would require significant rework/design on Editor side for how they do things. For now, I feel like this is at least a short-term solution that mitigates the problem.
@olegtk Do you have a perspective on this? Would Editor ever consider doing the rollback work to the starting point for command handlers instead of leaving the work up to individual languages?

**Review by Youssef1313 — COMMENTED** [review]
Does CompleteStatementCommandHandler (and potentially other command handlers) need a similar change?

**Review by sharwell — APPROVED** [review]

**Comment by allisonchou** [comment]
> Does CompleteStatementCommandHandler (and potentially other command handlers) need a similar change?

Hmm, I don't think so. The main issue in the problematic handlers was that `nextCommandHandler()` would be called, and that call could modify the text buffer, and then after that call we could throw cancellation in the handler, which violates the Editor API guidelines and could mess up the current state of things. `CompleteStatementCommandHandler` doesn't do any work that may trigger cancellation after `nextCommandHandler()` is called, so I think it's safe.

**Comment by allisonchou** [comment]
@CyrusNajmabadi Do you have any further concerns with this PR? I agree that the work to potentially rollback changes would ideally be done on the Editor side, however that seems like it would take quite a bit of planning and work on their end. At least for now, I feel like this PR is a good mitigation to the issues being observed.

**Comment by CyrusNajmabadi** [comment]
Effectively it feels fragile.  let me review again.

**Comment by allisonchou** [comment]
Talked with Cyrus/Sam offline - seems like this is OK to be merged as a temporary solution, however in the near future we'd like to make cancellation handling in command handlers more robust so that each handler does not need to deal with cancellation individually. This is tracked by #52528.

**Pull request merged**